catsoop/lti.py

```diff
@@ -0,0 +1,297 @@
+        if success:
+            LOGGER.info("[lti.lti4cs_response.send_outcome] outcome sent successfully")
+        else:
+            LOGGER.error("[lti.lti4cs_response.send_outcome] outcome sending FAILED")
+
+    def generate_result_xml(self, result_sourcedid, score):
+        '''
+        Create the XML document that contains the new score to be sent to the LTI
+        consumer. The format of this message is defined in the LTI 1.1 spec.
+        '''
+        elem = ElementMaker(nsmap={None: 'http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0'})
+        xml = elem.imsx_POXEnvelopeRequest(
+            elem.imsx_POXHeader(
+                elem.imsx_POXRequestHeaderInfo(
+                    elem.imsx_version('V1.0'),
+                    elem.imsx_messageIdentifier(str(uuid.uuid4()))
+                )
+            ),
+            elem.imsx_POXBody(
+                elem.replaceResultRequest(
+                    elem.resultRecord(
+                        elem.sourcedGUID(
+                            elem.sourcedId(result_sourcedid)
+                        ),
+                        elem.result(
+                            elem.resultScore(
+                                elem.language('en'),
+                                elem.textString(str(score))
+                            )
+                        )
+                    )
+                )
+            )
+        )
+        xml = etree.tostring(xml, xml_declaration=True, encoding='UTF-8')	# bytes
+        xml = xml.decode("utf-8")
+        return xml
+
+    @property
+    def response_url(self):
+        """
+        Returns remapped lis_outcome_service_url
+        uses pylti_url_fix map to support edX dev-stack
+
+        :return: remapped lis_outcome_service_url
+        """
+        url = ""
+        url = self.lti_data['lis_outcome_service_url']
+        urls = self.pylti_url_fix
+        # url remapping is useful for using edX devstack
+        # edX devstack reports httpS://localhost:8000/ and listens on HTTP
+        for prefix, mapping in urls.items():
+            if url.startswith(prefix):
+                for _from, _to in mapping.items():
+                    url = url.replace(_from, _to)
+        return url
+
+def get_or_create_user(context, uname, email, name):
+    '''
+    Ensure that the named user (by username) exists; create if necessary
+    '''
+    session_data = context["cs_session_data"]
+    logging = context["csm_cslog"]
+    LOGGER.info("[lti.get_or_create_user] uname=%s, email=%s, name=%s" % (uname, email, name))
+
+    login_info = logging.most_recent("_logininfo", [], uname, {})
+    if not login_info:	# account doesn't exist, create
+
+        LOGGER.info("[lti.get_or_create_user] uname=%s unknown username -> creating new account")
+        passwd = ''.join([random.choice(string.ascii_letters + string.digits) for n in range(32)])
+        hash_iterations = context.get("cs_password_hash_iterations", 500000)
```

Comment:
Just to make sure I'm understanding: we're never actually expecting someone to try to login with these credentials, right?  If not, we could probably just avoid the whole hashing process and just use a dummy (or empty) bytesstring as the password hash and salt in the new user's information (since it looks like we're just checking for the _presence_ of the user info, and not the contents).

If we do want someone, in theory, to be able to log in to these accounts using a password, there's some more work to be done to make sure that the password we generate here actually works (since there is some client-side hashing that happens in a normal login).

Reply:
This would supersede the comment below (about pulling those functions directly from the login question type).  Going even further, is there a reason even to bother with `_logininfo`?  We could force authentication through the session variables alone and avoid the need to create/check whether a regular account exists.

Reply:
Correct - the password & salt should never actually be used.  Forcing auth without regular account existence checking would be good - I was just avoiding modifications to your core auth code (no unit tests to ensure I didn't break things).

Reply:
OK, in that case I would forget about checking existence / creating accounts, and just set the session data appropriately.  That ends up being a lot simpler, and a lot faster since it won't have to hit the disk.

Reply:
Done - see https://github.com/ichuang/catsoop/pull/2/commits/6f4aaf7ccbf43ee893b26f974d11e18991897557 in https://github.com/ichuang/catsoop/pull/2